### ISSUE TITLE
Fragment selection skips namespace declaration for xsi:type value

### ISSUE THREAD

**Issue by ahenket** [eXist-db/exist#2182]
When you select a piece of XML , all relevant namespace declarations are copied into the fragment, except when the @xsi:type value has one. The net result is now an invalid fragment.

Tested with eXist-db 2.2 and 4.4.0 on macOS Mojave, you may run the following code to reproduce:

```
xquery version "3.0";

let $xml    :=
    <MCCI_IN200101 xmlns="urn:hl7-org:v3" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:hl7nl="urn:hl7-nl:v3">
        <effectiveTime xsi:type="hl7nl:PIVL_TS"/>
    </MCCI_IN200101>

return
    $xml/*
```

Actual result:

```<effectiveTime xmlns="urn:hl7-org:v3" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:type="hl7nl:PIVL_TS"/>```

Expected result:
```<effectiveTime xmlns="urn:hl7-org:v3" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:hl7nl="urn:hl7-nl:v3" xsi:type="hl7nl:PIVL_TS"/>```

**Comment by joewiz** [comment]
I'm genuinely unsure where in the spec the feature that would yield the expected outcome is stated or implied. Is it the static typing feature, https://www.w3.org/TR/xquery-31/#id-static-typing-feature?

For comparison:

- BaseX returns the same actual result as eXist.
- Saxon returns the expected result.

**Comment by duncdrum** [comment]
I m very much leaning to Saxon being right, and both eXist and BaseX being wrong. @adamretter thoughts?

**Comment by joewiz** [comment]
@ahenket Thank you for that reference. I actually meant "where in the *XQuery* spec". I'd like to be able to point to a section in the XQuery specification that says that processors must respect this. There are a number of schema- and type- related features that are labeled "optional" here: https://www.w3.org/TR/xquery-31/#id-conform-optional-features. And eXist's documentation says which optional features it doesn't support: http://exist-db.org/exist/apps/doc/xquery.xml#unsupported-features. 

Basically, I'm wondering whether this is a bug (failure to implement a required feature) or just an omission of an optional feature. 

To know how to classify this, we'll need to find the place in the XQuery specification that addresses this.

**Comment by ahenket** [comment]
I can only say that if you don't respect the QName(), you may end up with illegal XML. I don't see how you could skip implementation in that sense.

**Comment by duncdrum** [comment]
no bad query here, just a bug

**Comment by joewiz** [comment]
Duncan, do you see where in the spec the required behavior is described?

**Comment by ahenket** [comment]
In the mean time I have worked my way around it. This had to be done in two steps:

- Send full xml to XSLT that will extract the right path and copy required namespace declarations

That is is enough for schema validation, but not for schematron validation as upon transform:transform() eXist kick in and kills my carefully crafted namespace declarations again thus leading to incorrect errors

- Save the result of the XSLT to disk, and send doc(...)/* to transform:transform(), and clean up the temp file afterwards.

These are really nasty hacks that I'd appreciate to get rid of.

[add-namespace.xsl.txt](https://github.com/eXist-db/exist/files/2457925/add-namespace.xsl.txt)

**Comment by duncdrum** [comment]
@joewiz i d say thats here https://www.w3.org/TR/xquery-31/#id-ns-nodes-on-elements but I can't really see how the differences between basex/exist and saxon are covered there. 

One could argue that the following
```
Copy-namespaces mode does not affect the namespace bindings of a newly constructed element node. It applies only to existing nodes that are copied by a constructor expression.
```

implies `xmlns:hl7nl="urn:hl7-nl:v3"` can be dropped since it is not in use for the newly constructed element `$xml/*`, but the prefix should still be resolved imv (ala Saxon), hence i m leaning to bug.

**Comment by daliboris** [comment]
I have the similar issue, but with namespaces in the content, ie. attribute values, not attribute names.

```xquery
declare copy-namespaces preserve, inherit;

declare namespace t = "http://exist-db.org/xquery/tests/test";
declare namespace xs = "http://www.w3.org/2001/XMLSchema";
declare namespace tei = "http://www.tei-c.org/ns/1.0";
declare namespace cc = "http://exist-db.org/collection-config/1.0";

let $items :=  <t:items xmlns:t="http://exist-db.org/xquery/tests/test" xmlns="http://exist-db.org/collection-config/1.0" xmlns:tei="http://www.tei-c.org/ns/1.0" xmlns:xs="http://www.w3.org/2001/XMLSchema">
  <t:item>
    <collection>
      <index>
        <lucene>
          <text qname="tei:num">
            <field name="p-field" type="xs:integer"/>
          </text>
        </lucene>
      </index>
    </collection>
  </t:item>
</t:items>

let $item := $items/t:item/cc:collection

return $item
```
in eXist-db (6.4.0) returns

```xml
<collection xmlns="http://exist-db.org/collection-config/1.0">
 <index>
   <lucene><text qname="tei:num"><field name="p-field" type="xs:integer"/></text></lucene>
 </index>
</collection>
```

But the same query with

```xquery

declare copy-namespaces preserve, no-inherit;

```

returns **nothing**, ie. looks like it removes namespaces/namespace prefixes from the XPath, before serializing the output.

In Saxon, in both cases, the result is

```xml
<collection xmlns="http://exist-db.org/collection-config/1.0"
            xmlns:t="http://exist-db.org/xquery/tests/test"
            xmlns:tei="http://www.tei-c.org/ns/1.0"
            xmlns:xs="http://www.w3.org/2001/XMLSchema">
   <index>
      <lucene>
         <text qname="tei:num">
            <field name="p-field" type="xs:integer"/>
         </text>
      </lucene>
   </index>
</collection
```

I found this in Pciscila's XQuery, Second Edition, on `declare copy-namespaces` declaration (pages 174–175):

> The first setting, either `preserve` or `no-preserve`, controls whether unused namespace declarations are included from the input document. Unused here means “not used in the name of the element or its attributes.”

> One reason to choose `preserve` is if your element content (or attribute values) contains namespace-sensitive values, such as qualified names. Namespaces used in content (as opposed to in element/attribute names) are not considered to be “used.” If you choose no-preserve, and the prefixes in the content are dependent on those non-preserved declarations, an error may be raised when you try to validate or use the output document, because there may be an undeclared prefix.

> The second setting, either `inherit` or `no-inherit`, controls whether the in-scope namespace declarations are copied from an outer constructor in the query to elements that are being copied (usually from the input document). Choosing `inherit` usually results in a less cluttered result document. If this value is set to `no-inherit`, unused namespaces declared on ancestor elements are explicitly undeclared on the copied element. 

See [example 10-18](https://www.datypic.com/books/xquery/chapter10.html#example.Query-with-preserve-no-inherit)

**Comment by adamretter** [comment]
I think many of the previous comments in this thread have gone off in strange directions about namespaces, this isn't about namespaces, this issue IMHO is about XDM typing.

> In Saxon

@daliboris Which version of Saxon?

I think this is related to XML Schema support of typing in XQuery, Saxon-EE has that, but Saxon-PE doesn't, neither does eXist-db.

In this piece of XML:

```xml
<text qname="tei:num">
```

there is nothing special about the attribute name `qname`, it is just the name of an attribute, it does not infer any type information about its value.

In a non-schema aware XQuery processor, the type for the value of the `qname` attribute is `xs:untypedAtomic`. Therefore `tei:num` is not a QName and does not need a namespace (rather, for simplicities sake it can be though of as any other `xs:string` value).

In a schema-aware XQuery processor, the type for the value of the `qname` attribute may be resolved from an XML Schema (if resolvable), and at that point it will depend on the type specified in the XML Schema. This may or may not be an `xs:QName`.

If we look in `schemas/collection.xconf.xsd` we see that this is specified as an `xs:string` - this is a lazy oversimplification of what the Java code actually allows:
1. `xs:NCName`
2. `xs:QName`
3. `xs:string` (starting with a literal `@` prefix followed by `xs:NCName` or `xs:QName`)

Regardless, fixing the Schema won't help you as eXist-db does not support Schema Aware processing in XQuery.

So I don't think this is a bug per-se... maybe a large feature request to add Schema Aware processing ;-)

**Comment by daliboris** [comment]
Hello @adamretter.

> Which version of Saxon?

Both *expected* outputs are generated by Saxon HE-Query 12.9.

Sorry, I'm not an advanced expert in XML and I've never gotten deeply into XDM.

I my examples, I don't validate the output, I'm just wondering why in my first example not all namespaces and prefixes from the parent XML element are available in the output.

And in the second case: how can namespaces _disappear_ so that they cannot be used for XPath queries?
